VibrationActuator should stop vibrating when its document becomes hidden

cdumez opened the pull request:
#### 6d29a35a2b0da6ab9ae6756dd7eb262fc77a51a2
<pre>
VibrationActuator should stop vibrating when its document becomes hidden
<a href="https://bugs.webkit.org/show_bug.cgi?id=250414">https://bugs.webkit.org/show_bug.cgi?id=250414</a>

Reviewed by Youenn Fablet.

VibrationActuator should stop vibrating when its document becomes hidden:
- <a href="https://w3c.github.io/gamepad/extensions.html#gamepadhapticactuator-interface">https://w3c.github.io/gamepad/extensions.html#gamepadhapticactuator-interface</a> (Page Visibility section)

Also address test flakiness by properly clearing mock gamepads between tests.
We used to call MockGamepadProvider::clearMockGamepads() inside the WebContent
process even though they are stored in the UIProcess nowadays.

* LayoutTests/gamepad/gamepad-vibration-document-no-longer-fully-active.html: Added.
* LayoutTests/gamepad/gamepad-vibration-document-no-longer-fully-active-expected.txt: Added.
* LayoutTests/gamepad/gamepad-vibration-visibility-change-expected.txt: Added.
* LayoutTests/gamepad/gamepad-vibration-visibility-change.html: Added.
* Source/WebCore/Modules/gamepad/Gamepad.cpp:
(WebCore::Gamepad::Gamepad):
(WebCore::Gamepad::vibrationActuator):
* Source/WebCore/Modules/gamepad/Gamepad.h:
* Source/WebCore/Modules/gamepad/GamepadHapticActuator.cpp:
(WebCore::GamepadHapticActuator::create):
(WebCore::GamepadHapticActuator::GamepadHapticActuator):
(WebCore::GamepadHapticActuator::~GamepadHapticActuator):
(WebCore::GamepadHapticActuator::playEffect):
(WebCore::GamepadHapticActuator::reset):
(WebCore::GamepadHapticActuator::stopEffects):
(WebCore::GamepadHapticActuator::document):
(WebCore::GamepadHapticActuator::activeDOMObjectName const):
(WebCore::GamepadHapticActuator::suspend):
(WebCore::GamepadHapticActuator::stop):
(WebCore::GamepadHapticActuator::visibilityStateChanged):
* Source/WebCore/Modules/gamepad/GamepadHapticActuator.h:
* Source/WebCore/Modules/gamepad/GamepadHapticActuator.idl:
* Source/WebCore/Modules/gamepad/GamepadManager.cpp:
(WebCore::navigatorGamepadFromDOMWindow):
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::NavigatorGamepad):
(WebCore::NavigatorGamepad::from):
(WebCore::NavigatorGamepad::gamepadFromPlatformGamepad):
(WebCore::NavigatorGamepad::getGamepads):
(WebCore::NavigatorGamepad::gamepadsBecameVisible):
(WebCore::NavigatorGamepad::gamepadConnected):
* Source/WebCore/Modules/gamepad/NavigatorGamepad.h:
* Source/WebCore/Modules/webxr/WebXRInputSource.cpp:
(WebCore::WebXRInputSource::WebXRInputSource):
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::document):
* Source/WebCore/page/Navigator.h:
* Source/WebCore/testing/MockGamepad.cpp:
(WebCore::MockGamepad::MockGamepad):

Canonical link: <a href="https://commits.webkit.org/258802@main">https://commits.webkit.org/258802@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30c2a50eb42d635a6f260ee8c35471a5a44c0b70

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102971 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112221 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/172436 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106930 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13113 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2996 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95210 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/110291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108744 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37708 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/24805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5529 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/26214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5682 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2665 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11692 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45713 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6054 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7431 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/gamepad/GamepadHapticActuator.cpp:
```diff
@@ -91,33 +102,68 @@ void GamepadHapticActuator::playEffect(Document& document, EffectType effectType
+        return completionHandler();
+
+    queueTaskKeepingObjectAlive(*this, TaskSource::Gamepad, [playingEffectPromise = std::exchange(m_playingEffectPromise, nullptr)] {
+        playingEffectPromise->resolve<IDLEnumeration<Result>>(Result::Preempted);
```
This is not strictly as per spec as it seems the spec asks to set m_playingEffectPromise to undefined in the queued task, not when enqueuing the task.
I think the proposed patch behavior is good though, it might be worth filing a spec issue.

cdumez replied:
Yes, the spec has issues when it comes to this promise, tracked by https://github.com/w3c/gamepad/issues/179.

cdumez committed:
VibrationActuator should stop vibrating when its document becomes hidden
https://bugs.webkit.org/show_bug.cgi?id=250414

Reviewed by Youenn Fablet.

VibrationActuator should stop vibrating when its document becomes hidden:
- https://w3c.github.io/gamepad/extensions.html#gamepadhapticactuator-interface (Page Visibility section)

Also address test flakiness by properly clearing mock gamepads between tests.
We used to call MockGamepadProvider::clearMockGamepads() inside the WebContent
process even though they are stored in the UIProcess nowadays.

* LayoutTests/gamepad/gamepad-vibration-document-no-longer-fully-active.html: Added.
* LayoutTests/gamepad/gamepad-vibration-document-no-longer-fully-active-expected.txt: Added.
* LayoutTests/gamepad/gamepad-vibration-visibility-change-expected.txt: Added.
* LayoutTests/gamepad/gamepad-vibration-visibility-change.html: Added.
* Source/WebCore/Modules/gamepad/Gamepad.cpp:
(WebCore::Gamepad::Gamepad):
(WebCore::Gamepad::vibrationActuator):
* Source/WebCore/Modules/gamepad/Gamepad.h:
* Source/WebCore/Modules/gamepad/GamepadHapticActuator.cpp:
(WebCore::GamepadHapticActuator::create):
(WebCore::GamepadHapticActuator::GamepadHapticActuator):
(WebCore::GamepadHapticActuator::~GamepadHapticActuator):
(WebCore::GamepadHapticActuator::playEffect):
(WebCore::GamepadHapticActuator::reset):
(WebCore::GamepadHapticActuator::stopEffects):
(WebCore::GamepadHapticActuator::document):
(WebCore::GamepadHapticActuator::activeDOMObjectName const):
(WebCore::GamepadHapticActuator::suspend):
(WebCore::GamepadHapticActuator::stop):
(WebCore::GamepadHapticActuator::visibilityStateChanged):
* Source/WebCore/Modules/gamepad/GamepadHapticActuator.h:
* Source/WebCore/Modules/gamepad/GamepadHapticActuator.idl:
* Source/WebCore/Modules/gamepad/GamepadManager.cpp:
(WebCore::navigatorGamepadFromDOMWindow):
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::NavigatorGamepad):
(WebCore::NavigatorGamepad::from):
(WebCore::NavigatorGamepad::gamepadFromPlatformGamepad):
(WebCore::NavigatorGamepad::getGamepads):
(WebCore::NavigatorGamepad::gamepadsBecameVisible):
(WebCore::NavigatorGamepad::gamepadConnected):
* Source/WebCore/Modules/gamepad/NavigatorGamepad.h:
* Source/WebCore/Modules/webxr/WebXRInputSource.cpp:
(WebCore::WebXRInputSource::WebXRInputSource):
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::document):
* Source/WebCore/page/Navigator.h:
* Source/WebCore/testing/MockGamepad.cpp:
(WebCore::MockGamepad::MockGamepad):

Canonical link: https://commits.webkit.org/258802@main
